spec/03-plugins/25-oauth2/03-access_spec.lua

```diff
@@ -1373,6 +1374,7 @@ describe("Plugin: oauth2 [#" .. strategy .. "]", function()
           })
           local body = assert.res_status(200, res)
           assert.is_table(ngx.re.match(body, [[^\{"token_type":"bearer","access_token":"[\w]{32,32}","expires_in":5\}$]]))
+          assert.is_table(kong.client.get_consumer())
```

Comment:
Integration tests do not run inside of Kong, but alongside. Using the PDK in integration tests will never work. This is why the PDK is tested via unit tests in `t/`.

Reply:
Ah I see, looks like Kong has never been in a situation where a plugin needed a unit test in the ```/t``` directory w the pdk to confirm expected behavior, but it sounds like that is whats needed here. I supposed I would face the same problem trying to check confirmation of table ctx consumer via ```ngx.ctx.authenticated_consumer``` for the same reasons(run inside vs alongside) in the integration tests. 

So need something in ```/t``` that:

1. Stands up an oauth proxy and other needed Kong resources(consumer/consumer oauth creds/service/route/oauth2 plugin)

2. Calls the ```client_credentials``` grant token generation endpoint ```/oauth2/token``` with the consumer creds

3. Validates the pdk detects the consumers context set via ```kong.client.get_consumer```

Feels overkill for the minor improvement to oauth2 token gen existing behavior, but should be able to do it to maintain good visibility and keep the feature in a detectable state if somehow someone ever broke it. I just see no examples in the ```/t``` where Kong specific resources are being created and its not all just generic nginx checks so flying blind a bit. If there is any simpler approach do lemme know, was thinking to clean this up this weekend one evening among a few other pet projects I am tackling.

Reply:
For this we typically develop a tests plugin: https://github.com/Kong/kong/tree/master/spec/fixtures/custom_plugins/kong/plugins

Reply:
Cool, thanks for the link. There would not be much value if I have to rewrite the same logic in the oauth2 plugin in isolate(cause we want to hit the actual plugin, not dup code of it) but that might not be whats happening here. Did just try the ngx.ctx.auth_consumer field but fully expect this commit to fail before I try to see how the tests plugins relate to where they are implemented and validated and how it helps my use case here.

Reply:
To do this, we typically configure an OAuth plugin and one of those tests plugins onto a Route in the integration tests. The test plugin typically gets executed after the OAuth plugin and and plays the role of asserting that any `ngx.ctx` or other contextual values were properly set.

Reply:
To be clear when you refer to integration tests you mean https://github.com/Kong/kong/tree/master/spec/02-integration right?

Reply:
Those and plugins integration tests yes.